packages/mattermost-redux/src/actions/channel_categories.ts

```diff
@@ -34,28 +34,55 @@ import {$ID} from 'mattermost-redux/types/utilities';
 import {insertMultipleWithoutDuplicates, insertWithoutDuplicates, removeItem} from 'mattermost-redux/utils/array_utils';
 
 export function expandCategory(categoryId: string) {
-    return {
-        type: ChannelCategoryTypes.CATEGORY_EXPANDED,
-        data: categoryId,
-    };
+    return setCategoryCollapsed(categoryId, false);
 }
 
 export function collapseCategory(categoryId: string) {
-    return {
-        type: ChannelCategoryTypes.CATEGORY_COLLAPSED,
-        data: categoryId,
-    };
+    return setCategoryCollapsed(categoryId, true);
+}
+
+export function setCategoryCollapsed(categoryId: string, collapsed: boolean) {
+    return patchCategory(categoryId, {
+        collapsed,
+    });
 }
 
 export function setCategorySorting(categoryId: string, sorting: CategorySorting) {
-    return (dispatch: DispatchFunc, getState: GetStateFunc) => {
+    return patchCategory(categoryId, {
+        sorting,
+    });
+}
+
+export function patchCategory(categoryId: string, patch: Partial<ChannelCategory>): ActionFunc {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
         const state = getState();
-        const category = getCategory(state, categoryId);
+        const currentUserId = getCurrentUserId(state);
 
-        return dispatch(updateCategory({
+        const category = getCategory(state, categoryId);
+        const patchedCategory = {
             ...category,
-            sorting,
-        }));
+            ...patch,
+        };
+
+        dispatch({
+            type: ChannelCategoryTypes.RECEIVED_CATEGORY,
+            data: patchedCategory,
+        });
+
+        try {
+            await Client4.updateChannelCategory(currentUserId, category.team_id, patchedCategory);
+        } catch (error) {
+            dispatch({
+                type: ChannelCategoryTypes.RECEIVED_CATEGORY,
+                data: category,
+            });
+
+            forceLogoutIfNecessary(error, dispatch, getState);
```

Comment:
would this be necessary if we get a category collapse error?

Reply:
All API requests actually have this in the error handler. It logs the app out if we receive a 401 error from the server since that means the user's session expired or was revoked. Although now that I say it's everywhere, it would be nice to have some sort of middleware handling this...